replica/database.hh

```diff
@@ -618,6 +618,12 @@ private:
     // can wait on compaction (backpressure) which in turn takes deletion guard on completion.
     future<> safe_foreach_sstable(const sstables::sstable_set&, noncopyable_function<future<>(const sstables::shared_sstable&)> action);
 
+    // Returns a sstable set that can be safely used for purging any expired tombstone in a compaction group.
+    // Only the sstables in the compaction group is not sufficient, since there might be other compaction
+    // groups during tablet split with overlapping token range, and we need to include them all in a single
+    // sstable set to allow safe tombstone gc.
+    lw_shared_ptr<const sstables::sstable_set> sstable_set_for_tombstone_gc(const compaction_group&) const;
```

Comment:
Does this affect all compactions?
Wouldn't it be simpler to just disable tombstone gc for split compaction?

Reply:
The contract between compaction and replica is: give me all sstables that possibly contain shadowed data, which means all that fall into the same token range. 

It's not enough to disable tombstone gc during split compaction, since 1) it's disabled already. 2) the problem is that sstables with overlapping token range can be temporarily segregated into different groups, so regular compaction in one of the groups is sufficient to cause the problem.

I cannot disable tombstone GC completely either (during split), since it can cause other problems since split can take long.

When the context is not split, then we just feed the sstable set from the single compaction group (so semantic remains the same). So this solution is about respecting the contract, by feeding all sstables that will allow tombstone GC decision to be correct.

Reply:
This is also a good solution looking forward to merge, where merge is completed already, but compaction groups (that previously belong'd to different tablets) will be merged asynchronously.

Reply:
@denesb I think there might be a misunderstanding here too. This sstable set is not used for deciding what to be compacted, it's only used for the shadowing check in tombstone GC.

Reply:
Ok, thanks for explaining.